builder/dockerfile/instructions/parse.go

```diff
@@ -0,0 +1,652 @@
+	if len(req.args) < 2 {
+		return nil, errAtLeastTwoArguments("COPY")
+	}
+
+	flFrom := req.flags.AddString("from", "")
+	if err := req.flags.Parse(); err != nil {
+		return nil, err
+	}
+	return &CopyCommand{
+		Srcs:        req.args[:len(req.args)-1],
+		Dest:        req.args[len(req.args)-1],
+		From:        flFrom.Value,
+		BaseCommand: newBaseCommand(req),
+	}, nil
+}
+
+func parseFrom(req parseRequest) (*FromCommand, error) {
+	// HELP HERE <- do we need to validate stageName at dispatch time (do we authorize env / buildArgs in here?)
+	stageName, err := parseBuildStageName(req.args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := req.flags.Parse(); err != nil {
+		return nil, err
+	}
+
+	return &FromCommand{
+		BaseName:    req.args[0],
+		StageName:   stageName,
+		BaseCommand: newBaseCommand(req),
+	}, nil
+
+	//TODO
+	//req.buildArgs.ResetAllowed()
+
+}
+
+func parseBuildStageName(args []string) (string, error) {
+	stageName := ""
+	switch {
+	case len(args) == 3 && strings.EqualFold(args[1], "as"):
+		stageName = strings.ToLower(args[2])
+		if ok, _ := regexp.MatchString("^[a-z][a-z0-9-_\\.]*$", stageName); !ok {
+			return "", errors.Errorf("invalid name for build stage: %q, name can't start with a number or contain symbols", stageName)
+		}
+	case len(args) != 1:
+		return "", errors.New("FROM requires either one or three arguments")
+	}
+
+	return stageName, nil
+}
+
+func parseOnBuild(req parseRequest) (*OnbuildCommand, error) {
+	if len(req.args) == 0 {
+		return nil, errAtLeastOneArgument("ONBUILD")
+	}
+	if err := req.flags.Parse(); err != nil {
+		return nil, err
+	}
+
+	triggerInstruction := strings.ToUpper(strings.TrimSpace(req.args[0]))
+	switch strings.ToUpper(triggerInstruction) {
+	case "ONBUILD":
+		return nil, errors.New("Chaining ONBUILD via `ONBUILD ONBUILD` isn't allowed")
+	case "MAINTAINER", "FROM":
+		return nil, fmt.Errorf("%s isn't allowed as an ONBUILD trigger", triggerInstruction)
+	}
+
+	original := regexp.MustCompile(`(?i)^\s*ONBUILD\s*`).ReplaceAllString(req.original, "")
+	return &OnbuildCommand{
+		Expression:  original,
+		BaseCommand: newBaseCommand(req),
+	}, nil
+
+}
+
+func parseWorkdir(req parseRequest) (*WorkdirCommand, error) {
+	if len(req.args) != 1 {
+		return nil, errExactlyOneArgument("WORKDIR")
+	}
+
+	err := req.flags.Parse()
+	if err != nil {
+		return nil, err
+	}
+	return &WorkdirCommand{
+		Path:        req.args[0],
+		BaseCommand: newBaseCommand(req),
+	}, nil
+
+}
+
+func parseRun(req parseRequest) (*RunCommand, error) {
+
+	if err := req.flags.Parse(); err != nil {
+		return nil, err
+	}
+	prependShell := false
+	args := handleJSONArgs(req.args, req.attributes)
+	if !req.attributes["json"] {
+		prependShell = true
+	}
+	cmdFromArgs := strslice.StrSlice(args)
+
+	return &RunCommand{
+		Expression:   cmdFromArgs,
+		PrependShell: prependShell,
+		BaseCommand:  newBaseCommand(req),
+	}, nil
+
+}
+
+func parseCmd(req parseRequest) (*CmdCommand, error) {
```

Comment:
My point is if we add flags to `cmd` in the future that are not in `run`, it would be easier if the code is already split (even if in its implementation it calls a single shared function with `run` and `entrypoint`). It is also more explicit: if I'm looking to change `run` parsing behavior, I know that I can modify parseRun without any side effect on other command parsing logic